File: backend/old/src/services/emojiService.ts
/**
 * Emoji表专用操作服务
 * 提供对emoji表的高级操作功能
 */

import { logger } from '../core/logger';
import { databaseManager } from '../core/database';
import { pathCacheManager } from '../core/pathCacheManager';
import { 
  EmojiRecord, 
  EmojiInsertData, 
  EmojiUpdateData,
  EmojiQueryFilter,
  EmojiPaginationParams,
  EmojiQueryResponse
} from '../types/emoji';
import { QueryFilter, PaginationParams, SortParams } from '../types/database';
import { ValidationError, NotFoundError, InternalServerError } from '../middleware/errorHandler';
import path from 'path';
import fs from 'fs';
import crypto from 'crypto';

export class EmojiService {
  private static readonly TABLE_NAME = 'emoji';
  private static readonly CONNECTION_NAME = 'maibot';

  /**
   * 初始化麦麦数据库连接
   */
  private static async initConnection(): Promise<void> {
    const mainRoot = pathCacheManager.getMainRoot();
    if (!mainRoot) {
      throw new NotFoundError('麦麦主程序根目录未设置，请先设置根目录缓存');
    }

    const dbPath = path.join(mainRoot, 'data', 'MaiBot.db');
    
    // 检查连接是否已存在
    if (!databaseManager.hasConnection(this.CONNECTION_NAME)) {
      if (!databaseManager.isManagerInitialized()) {
        await databaseManager.initialize();
      }
      
      await databaseManager.addDatabase(this.CONNECTION_NAME, {
        path: dbPath,
        readonly: false,
        timeout: 5000,
        fileMustExist: true
      });
      
      logger.info('麦麦数据库连接已建立');
    }
  }

  /**
   * 验证emoji数据
   */
  private static validateEmojiData(data: EmojiInsertData | EmojiUpdateData): void {
    const requiredFields = ['full_path', 'format', 'emoji_hash', 'description', 'emotion'];
    
    for (const field of requiredFields) {
      if (!(field in data) || !data[field as keyof typeof data]) {
        throw new ValidationError(`字段 ${field} 不能为空`);
      }
    }

    // 验证数字字段
    const numericFields = ['query_count', 'is_registered', 'is_banned', 'record_time', 'register_time', 'usage_count', 'last_used_time'];
    for (const field of numericFields) {
      if (field in data && typeof data[field as keyof typeof data] !== 'number') {
        throw new ValidationError(`字段 ${field} 必须是数字类型`);
      }
    }

    // 验证布尔字段（0或1）
    if ('is_registered' in data && (data.is_registered !== 0 && data.is_registered !== 1)) {
      throw new ValidationError('is_registered 必须是0或1');
    }

    if ('is_banned' in data && (data.is_banned !== 0 && data.is_banned !== 1)) {
      throw new ValidationError('is_banned 必须是0或1');
    }
  }

  /**
   * 构建查询条件
   */
  private static buildWhereConditions(filter?: EmojiQueryFilter): QueryFilter[] {
    if (!filter) return [];

    const conditions: QueryFilter[] = [];

    if (filter.format) {
      conditions.push({ field: 'format', operator: '=', value: filter.format });
    }

    if (filter.emotion) {
      conditions.push({ field: 'emotion', operator: '=', value: filter.emotion });
    }

    if (filter.is_registered !== undefined) {
      conditions.push({ field: 'is_registered', operator: '=', value: filter.is_registered });
    }

    if (filter.is_banned !== undefined) {
      conditions.push({ field: 'is_banned', operator: '=', value: filter.is_banned });
    }

    if (filter.description) {
      conditions.push({ field: 'description', operator: 'LIKE', value: `%${filter.description}%` });
    }

    if (filter.emoji_hash) {
      conditions.push({ field: 'emoji_hash', operator: '=', value: filter.emoji_hash });
    }

    return conditions;
  }

  /**
   * 查询emoji（分页）
   */
  public static async getEmojis(params: EmojiPaginationParams): Promise<EmojiQueryResponse> {
    try {
      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 验证分页参数
      if (params.page < 1) {
        throw new ValidationError('页码必须大于0');
      }

      if (params.pageSize < 1 || params.pageSize > 1000) {
        throw new ValidationError('页大小必须在1-1000之间');
      }

      // 构建查询参数
      const paginationParams: PaginationParams = {
        page: params.page,
        pageSize: params.pageSize,
        offset: (params.page - 1) * params.pageSize
      };

      const queryParams: any = {};

      // 构建WHERE条件
      if (params.filter) {
        queryParams.where = this.buildWhereConditions(params.filter);
      }

      // 构建排序条件
      if (params.orderBy) {
        const orderBy: SortParams[] = [{
          field: params.orderBy,
          order: params.orderDir || 'ASC'
        }];
        queryParams.orderBy = orderBy;
      } else {
        // 默认按ID降序排列
        queryParams.orderBy = [{ field: 'id', order: 'DESC' }];
      }

      // 执行查询
      const result = await operator.findWithPagination<EmojiRecord>(
        this.TABLE_NAME,
        paginationParams,
        queryParams
      );

      return {
        items: result.data,
        totalPages: result.pagination.totalPages,
        currentPage: result.pagination.page,
        pageSize: result.pagination.pageSize,
        total: result.pagination.total,
        hasNext: result.pagination.hasNext,
        hasPrev: result.pagination.hasPrev
      };

    } catch (error) {
      logger.error('查询emoji失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError 
        ? error 
        : new InternalServerError('查询emoji失败');
    }
  }

  /**
   * 根据ID查询单个emoji
   */
  public static async getEmojiById(id: number): Promise<EmojiRecord | null> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.findOne<EmojiRecord>(this.TABLE_NAME, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      return result;

    } catch (error) {
      logger.error('根据ID查询emoji失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询emoji失败');
    }
  }

  /**
   * 插入emoji
   */
  public static async insertEmoji(data: EmojiInsertData): Promise<number> {
    try {
      this.validateEmojiData(data);

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 设置默认值（只对用户未提供的字段设置默认值）
      const currentTime = Date.now() / 1000;
      const defaults = {
        query_count: 0,
        is_registered: 0,
        is_banned: 0,
        record_time: currentTime,
        register_time: currentTime,
        usage_count: 0,
        last_used_time: currentTime,
      };
      
      // 合并数据，用户提供的数据优先
      const insertData = { ...defaults, ...data };

      const result = await operator.insertOne(this.TABLE_NAME, insertData);

      if (!result.lastInsertId) {
        throw new InternalServerError('插入emoji失败，未获得插入ID');
      }

      logger.info(`emoji插入成功，ID: ${result.lastInsertId}`);
      return result.lastInsertId;

    } catch (error) {
      logger.error('插入emoji失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('插入emoji失败');
    }
  }

  /**
   * 更新emoji
   */
  public static async updateEmoji(data: EmojiUpdateData): Promise<boolean> {
    try {
      if (!data.id || data.id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      // 验证更新数据（排除id字段）
      const { id, ...updateData } = data;
      if (Object.keys(updateData).length === 0) {
        throw new ValidationError('更新数据不能为空');
      }

      this.validateEmojiData(updateData as EmojiInsertData);

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 检查记录是否存在
      const existingRecord = await this.getEmojiById(id);
      if (!existingRecord) {
        throw new NotFoundError('要更新的emoji记录不存在');
      }

      const result = await operator.update(this.TABLE_NAME, updateData, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      if (result.affectedRows === 0) {
        throw new InternalServerError('更新emoji失败，未影响任何记录');
      }

      logger.info(`emoji更新成功，ID: ${id}`);
      return true;

    } catch (error) {
      logger.error('更新emoji失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('更新emoji失败');
    }
  }

  /**
   * 删除emoji
   */
  public static async deleteEmoji(id: number): Promise<boolean> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 检查记录是否存在
      const existingRecord = await this.getEmojiById(id);
      if (!existingRecord) {
        throw new NotFoundError('要删除的emoji记录不存在');
      }

      const result = await operator.delete(this.TABLE_NAME, {
        where: [{ field: 'id', operator: '=', value: id }]
      });

      if (result.affectedRows === 0) {
        throw new InternalServerError('删除emoji失败，未影响任何记录');
      }

      logger.info(`emoji删除成功，ID: ${id}`);
      return true;

    } catch (error) {
      logger.error('删除emoji失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('删除emoji失败');
    }
  }

  /**
   * 增加查询次数
   */
  public static async incrementQueryCount(id: number): Promise<void> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 使用原生SQL更新查询次数
      await operator.executeStatement(
        'UPDATE emoji SET query_count = query_count + 1, last_used_time = ? WHERE id = ?',
        [Date.now() / 1000, id]
      );

    } catch (error) {
      logger.error('增加查询次数失败:', error);
      // 这个操作失败不应该影响主要业务流程，只记录错误
    }
  }

  /**
   * 根据哈希值查询emoji
   */
  public static async getEmojiByHash(emojiHash: string): Promise<EmojiRecord | null> {
    try {
      if (!emojiHash) {
        throw new ValidationError('emoji哈希值不能为空');
      }

      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      const result = await operator.findOne<EmojiRecord>(this.TABLE_NAME, {
        where: [{ field: 'emoji_hash', operator: '=', value: emojiHash }]
      });

      return result;

    } catch (error) {
      logger.error('根据哈希值查询emoji失败:', error);
      throw error instanceof ValidationError 
        ? error 
        : new InternalServerError('查询emoji失败');
    }
  }

  /**
   * 根据ID获取emoji图片的Base64编码
   */
  public static async getEmojiImage(id: number): Promise<string> {
    try {
      if (!id || id < 1) {
        throw new ValidationError('ID必须是大于0的数字');
      }

      // 获取emoji记录
      const emojiRecord = await this.getEmojiById(id);
      if (!emojiRecord) {
        throw new NotFoundError(`未找到ID为 ${id} 的emoji记录`);
      }

      // 获取麦麦根目录
      const mainRoot = pathCacheManager.getMainRoot();
      if (!mainRoot) {
        throw new NotFoundError('麦麦主程序根目录未设置，请先设置根目录缓存');
      }

      // 构建完整文件路径
      // full_path 是相对于麦麦根目录的路径，例如：data\emoji_registed\1753422909_5df38f6a.gif
      const fullFilePath = path.join(mainRoot, emojiRecord.full_path);
      
      // 检查文件是否存在
      if (!fs.existsSync(fullFilePath)) {
        throw new NotFoundError(`emoji文件不存在: ${fullFilePath}`);
      }

      // 读取文件内容并转换为Base64
      try {
        const fileBuffer = fs.readFileSync(fullFilePath);
        const base64String = fileBuffer.toString('base64');
        
        logger.info(`成功读取emoji图片，ID: ${id}, 文件大小: ${fileBuffer.length} 字节`);
        return base64String;
      } catch (fileError) {
        logger.error(`读取emoji文件失败: ${fullFilePath}`, fileError);
        throw new InternalServerError(`读取emoji文件失败: ${fileError instanceof Error ? fileError.message : '未知错误'}`);
      }

    } catch (error) {
      logger.error('获取emoji图片失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('获取emoji图片失败');
    }
  }

  /**
   * 获取emoji统计信息
   */
  public static async getEmojiStats(): Promise<{
    total: number;
    registered: number;
    banned: number;
    byFormat: Record<string, number>;
    byEmotion: Record<string, number>;
  }> {
    try {
      await this.initConnection();
      const operator = databaseManager.getOperator(this.CONNECTION_NAME);

      // 获取总数
      const total = await operator.count(this.TABLE_NAME);

      // 获取已注册数量
      const registered = await operator.count(this.TABLE_NAME, {
        where: [{ field: 'is_registered', operator: '=', value: 1 }]
      });

      // 获取被禁止数量
      const banned = await operator.count(this.TABLE_NAME, {
        where: [{ field: 'is_banned', operator: '=', value: 1 }]
      });

      // 按格式统计
      const formatStats = await operator.executeQuery<{ format: string; count: number }>(
        'SELECT format, COUNT(*) as count FROM emoji GROUP BY format ORDER BY count DESC'
      );

      // 按情感统计
      const emotionStats = await operator.executeQuery<{ emotion: string; count: number }>(
        'SELECT emotion, COUNT(*) as count FROM emoji GROUP BY emotion ORDER BY count DESC'
      );

      return {
        total,
        registered,
        banned,
        byFormat: Object.fromEntries(formatStats.map(item => [item.format, item.count])),
        byEmotion: Object.fromEntries(emotionStats.map(item => [item.emotion, item.count]))
      };

    } catch (error) {
      logger.error('获取emoji统计信息失败:', error);
      throw new InternalServerError('获取emoji统计信息失败');
    }
  }

  /**
   * 根据图片路径计算哈希值
   */
  public static async calculateImageHash(imagePath: string): Promise<string> {
    try {
      if (!imagePath) {
        throw new ValidationError('图片路径不能为空');
      }

      // 获取麦麦根目录
      const mainRoot = pathCacheManager.getMainRoot();
      if (!mainRoot) {
        throw new NotFoundError('麦麦主程序根目录未设置，请先设置根目录缓存');
      }

      // 构建完整文件路径
      const fullFilePath = path.join(mainRoot, imagePath);
      
      // 检查文件是否存在
      if (!fs.existsSync(fullFilePath)) {
        throw new NotFoundError(`图片文件不存在: ${fullFilePath}`);
      }

      // 读取文件内容
      try {
        const fileBuffer = fs.readFileSync(fullFilePath);
        
        // 计算MD5哈希值 (与第462行的算法保持一致)
        const imageHash = crypto.createHash('md5').update(fileBuffer).digest('hex');
        
        logger.info(`成功计算图片哈希值，路径: ${imagePath}, 哈希值: ${imageHash}, 文件大小: ${fileBuffer.length} 字节`);
        return imageHash;
        
      } catch (fileError) {
        logger.error(`读取图片文件失败: ${fullFilePath}`, fileError);
        throw new InternalServerError(`读取图片文件失败: ${fileError instanceof Error ? fileError.message : '未知错误'}`);
      }

    } catch (error) {
      logger.error('计算图片哈希值失败:', error);
      throw error instanceof ValidationError || error instanceof NotFoundError
        ? error 
        : new InternalServerError('计算图片哈希值失败');
    }
  }
}

File: backend/src/services/emoji_service.py
"""
Emoji表专用操作服务
提供对emoji表的高级操作功能
"""

import os
import hashlib
import base64
from typing import Optional, Dict, Any
from models.emoji import (
    EmojiRecord, EmojiInsertData, EmojiUpdateData, 
    EmojiQueryResponse, EmojiPaginationParams
)
from core.database_manager import database_manager
from core.path_cache_manager import path_cache_manager
from models.database import OrderDirection
import logging

logger = logging.getLogger("HMML")


class EmojiService:
    """Emoji表操作服务"""
    
    TABLE_NAME = "emoji"
    CONNECTION_NAME = "maibot"
    
    @classmethod
    async def _get_operator(cls):
        """获取数据库操作器"""
        operator = database_manager.get_operator(cls.CONNECTION_NAME)
        if not operator:
            # 尝试重新初始化麦麦数据库连接
            await database_manager._initialize_maimai_database()
            operator = database_manager.get_operator(cls.CONNECTION_NAME)
            
        if not operator:
            raise RuntimeError("麦麦数据库连接不可用，请检查数据库配置和路径缓存设置")
        return operator
        
    @classmethod
    async def get_emojis(cls, params: EmojiPaginationParams) -> EmojiQueryResponse:
        """
        分页查询emoji列表
        
        Args:
            params: 分页查询参数
            
        Returns:
            查询响应
        """
        try:
            operator = await cls._get_operator()
            
            # 构建WHERE条件
            where_conditions = {}
            if params.format:
                where_conditions["format"] = params.format
            if params.emotion:
                where_conditions["emotion"] = f"%{params.emotion}%"
            if params.description:
                where_conditions["description"] = f"%{params.description}%"
            if params.emoji_hash:
                where_conditions["emoji_hash"] = params.emoji_hash
            if params.is_registered is not None:
                where_conditions["is_registered"] = params.is_registered
            if params.is_banned is not None:
                where_conditions["is_banned"] = params.is_banned
                
            # 验证排序字段
            valid_order_fields = ['id', 'query_count', 'usage_count', 'last_used_time', 'record_time']
            order_by = None
            order_dir = OrderDirection.ASC
            
            if params.order_by and params.order_by in valid_order_fields:
                order_by = params.order_by
                if params.order_dir and params.order_dir.upper() == "DESC":
                    order_dir = OrderDirection.DESC
                    
            # 执行分页查询
            result = operator.find_with_pagination(
                table_name=cls.TABLE_NAME,
                page=params.page,
                page_size=params.page_size,
                where_conditions=where_conditions if where_conditions else None,
                order_by=order_by,
                order_dir=order_dir
            )
            
            # 转换为emoji记录
            emoji_items = []
            for item in result.items:
                emoji_items.append(EmojiRecord(**item))
                
            return EmojiQueryResponse(
                items=emoji_items,
                total_pages=result.total_pages,
                current_page=result.current_page,
                page_size=result.page_size,
                total=result.total,
                has_next=result.has_next,
                has_prev=result.has_prev
            )
            
        except Exception as error:
            logger.error(f"查询emoji列表失败: {error}")
            raise
            
    @classmethod
    async def get_emoji_by_id(cls, emoji_id: int) -> Optional[EmojiRecord]:
        """
        根据ID查询单个emoji
        
        Args:
            emoji_id: emoji ID
            
        Returns:
            emoji记录或None
        """
        try:
            if emoji_id <= 0:
                raise ValueError("emoji ID必须大于0")
                
            operator = await cls._get_operator()
            
            result = operator.find_one(
                table_name=cls.TABLE_NAME,
                where_conditions={"id": emoji_id}
            )
            
            if result:
                return EmojiRecord(**result)
            return None
            
        except Exception as error:
            logger.error(f"根据ID查询emoji失败: {error}")
            raise
            
    @classmethod
    async def insert_emoji(cls, data: EmojiInsertData) -> int:
        """
        插入emoji记录
        
        Args:
            data: 插入数据
            
        Returns:
            插入的记录ID
        """
        try:
            operator = await cls._get_operator()
            
            # 转换为字典格式
            insert_data = data.model_dump()
            
            # 执行插入
            result = operator.insert(cls.TABLE_NAME, insert_data)
            
            if not result.success or result.last_insert_id is None:
                raise RuntimeError("插入emoji失败")
                
            logger.info(f"emoji插入成功，ID: {result.last_insert_id}")
            return result.last_insert_id
            
        except Exception as error:
            logger.error(f"插入emoji失败: {error}")
            raise
            
    @classmethod
    async def update_emoji(cls, data: EmojiUpdateData) -> bool:
        """
        更新emoji记录
        
        Args:
            data: 更新数据
            
        Returns:
            是否更新成功
        """
        try:
            if data.id <= 0:
                raise ValueError("emoji ID必须大于0")
                
            operator = await cls._get_operator()
            
            # 构建更新数据（排除None值）
            update_data = {}
            for field, value in data.model_dump(exclude={'id'}).items():
                if value is not None:
                    update_data[field] = value
                    
            if not update_data:
                logger.warning("没有需要更新的数据")
                return False
                
            # 执行更新
            result = operator.update(
                table_name=cls.TABLE_NAME,
                data=update_data,
                where_conditions={"id": data.id}
            )
            
            if result.affected_rows > 0:
                logger.info(f"emoji更新成功，ID: {data.id}")
                return True
            else:
                logger.warning(f"未找到要更新的emoji，ID: {data.id}")
                return False
                
        except Exception as error:
            logger.error(f"更新emoji失败: {error}")
            raise
            
    @classmethod
    async def delete_emoji(cls, emoji_id: int) -> bool:
        """
        删除emoji记录
        
        Args:
            emoji_id: emoji ID
            
        Returns:
            是否删除成功
        """
        try:
            if emoji_id <= 0:
                raise ValueError("emoji ID必须大于0")
                
            operator = await cls._get_operator()
            
            # 执行删除
            result = operator.delete(
                table_name=cls.TABLE_NAME,
                where_conditions={"id": emoji_id}
            )
            
            if result.affected_rows > 0:
                logger.info(f"emoji删除成功，ID: {emoji_id}")
                return True
            else:
                logger.warning(f"未找到要删除的emoji，ID: {emoji_id}")
                return False
                
        except Exception as error:
            logger.error(f"删除emoji失败: {error}")
            raise
            
    @classmethod
    async def get_emoji_by_hash(cls, emoji_hash: str) -> Optional[EmojiRecord]:
        """
        根据哈希值查询emoji
        
        Args:
            emoji_hash: emoji哈希值
            
        Returns:
            emoji记录或None
        """
        try:
            if not emoji_hash:
                raise ValueError("emoji哈希值不能为空")
                
            operator = await cls._get_operator()
            
            result = operator.find_one(
                table_name=cls.TABLE_NAME,
                where_conditions={"emoji_hash": emoji_hash}
            )
            
            if result:
                return EmojiRecord(**result)
            return None
            
        except Exception as error:
            logger.error(f"根据哈希值查询emoji失败: {error}")
            raise
            
    @classmethod
    async def increment_query_count(cls, emoji_id: int) -> bool:
        """
        增加emoji查询次数
        
        Args:
            emoji_id: emoji ID
            
        Returns:
            是否更新成功
        """
        try:
            operator = await cls._get_operator()
            
            # 使用原始SQL更新查询次数
            sql = f"UPDATE {cls.TABLE_NAME} SET query_count = query_count + 1 WHERE id = ?"
            cursor = operator.execute_raw_sql(sql, (emoji_id,))
            
            # 提交事务
            operator.connection.get_connection().commit()
            
            return cursor.rowcount > 0
            
        except Exception as error:
            logger.error(f"增加emoji查询次数失败: {error}")
            raise
            
    @classmethod
    async def get_emoji_stats(cls) -> Dict[str, Any]:
        """
        获取emoji统计信息
        
        Returns:
            统计信息字典
        """
        try:
            operator = await cls._get_operator()
            
            # 总数统计
            total_count = operator.count(cls.TABLE_NAME)
            
            # 按格式统计
            cursor = operator.execute_raw_sql(
                f"SELECT format, COUNT(*) as count FROM {cls.TABLE_NAME} GROUP BY format"
            )
            format_stats = {row['format']: row['count'] for row in cursor.fetchall()}
            
            # 按注册状态统计
            registered_count = operator.count(cls.TABLE_NAME, {"is_registered": 1})
            unregistered_count = total_count - registered_count
            
            # 按禁用状态统计
            banned_count = operator.count(cls.TABLE_NAME, {"is_banned": 1})
            active_count = total_count - banned_count
            
            return {
                "total_count": total_count,
                "format_stats": format_stats,
                "registered_count": registered_count,
                "unregistered_count": unregistered_count,
                "banned_count": banned_count,
                "active_count": active_count
            }
            
        except Exception as error:
            logger.error(f"获取emoji统计信息失败: {error}")
            raise
            
    @classmethod
    async def get_emoji_image(cls, emoji_id: int) -> str:
        """
        获取emoji图片的Base64编码
        
        Args:
            emoji_id: emoji ID
            
        Returns:
            Base64编码的图片数据
        """
        try:
            if emoji_id <= 0:
                raise ValueError("emoji ID必须大于0")
                
            # 获取emoji记录
            emoji = await cls.get_emoji_by_id(emoji_id)
            if not emoji:
                raise RuntimeError(f"未找到ID为 {emoji_id} 的emoji记录")
                
            # 获取麦麦根目录
            main_root = path_cache_manager.get_main_root()
            if not main_root:
                raise RuntimeError("麦麦主程序根目录未设置")
                
            # 构建完整文件路径
            full_path = os.path.join(main_root, emoji.full_path.lstrip('/\\'))
            
            # 检查文件是否存在
            if not os.path.exists(full_path):
                raise RuntimeError(f"emoji图片文件不存在: {full_path}")
                
            # 读取文件并转换为Base64
            with open(full_path, 'rb') as f:
                file_data = f.read()
                base64_data = base64.b64encode(file_data).decode('utf-8')
                
            return base64_data
            
        except Exception as error:
            logger.error(f"获取emoji图片失败: {error}")
            raise
            
    @classmethod
    async def calculate_image_hash(cls, image_path: str) -> str:
        """
        计算图片文件的哈希值
        
        Args:
            image_path: 图片相对路径
            
        Returns:
            图片哈希值
        """
        try:
            if not image_path:
                raise ValueError("图片路径不能为空")
                
            # 获取麦麦根目录
            main_root = path_cache_manager.get_main_root()
            if not main_root:
                raise RuntimeError("麦麦主程序根目录未设置")
                
            # 构建完整文件路径
            full_path = os.path.join(main_root, image_path.lstrip('/\\'))
            
            # 检查文件是否存在
            if not os.path.exists(full_path):
                raise RuntimeError(f"图片文件不存在: {full_path}")
                
            # 计算文件MD5哈希值
            with open(full_path, 'rb') as f:
                file_data = f.read()
                hash_md5 = hashlib.md5(file_data)
                return hash_md5.hexdigest()
                
        except Exception as error:
            logger.error(f"计算图片哈希值失败: {error}")
            raise
